core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -135,7 +135,7 @@ private[transaction] class TransactionMetadata(val producerId: Long,
   }
 
   def removePartition(topicPartition: TopicPartition): Unit = {
-    if (pendingState.isDefined || (state != PrepareCommit && state != PrepareAbort))
+    if (pendingState.isDefined && (state != PrepareCommit && state != PrepareAbort))
```

Comment:
@guozhangwang does this look right to you? The basic integration test would hit this exception from the `WriteTxnMarkerRequestCompletionHandler` otherwise.